src/htbl.ml

```diff
@@ -0,0 +1,354 @@
+    let x = ((x land 0xcc_cc_cc_cc) lsr 2) lor ((x land 0x33_33_33_33) lsl 2) in
+    let x = ((x land 0xf0_f0_f0_f0) lsr 4) lor ((x land 0x0f_0f_0f_0f) lsl 4) in
+    let x = ((x land 0xff_00_ff_00) lsr 8) lor ((x land 0x00_ff_00_ff) lsl 8) in
+    (x lsr 16) lor (x lsl 16) land 0xffffffff
+
+  let compute_hkey k = reverse k lor 0x00_00_00_01
+
+  (** unset most significant turn on bit (for int32) *)
+  let unset_msb key =
+    let a = key lor (key lsr 1) in
+    let a = a lor (a lsr 2) in
+    let a = a lor (a lsr 4) in
+    let a = a lor (a lsr 8) in
+    let a = a lor (a lsr 16) in
+    (a lsr 1) land key
+
+  let new_dummy_node id llist =
+    let _, local = Llist.unsafe_add (reverse id) Dummy llist in
+    (* If the insertion succeeded ([is_added] = true) then
+       [local.prev] contains the atomic value of the new node.
+
+       If it failed that means another domain has already inserted
+       this dummy node. In this case, [local.prev] also contains the
+       value we seek as the [Llist.find] function calls by
+       [Llist.unsafe_add] will have stopped at the right place. *)
+    Atomic.get local.prev
+end
+
+module Htbl = struct
+  open Common
+  include Type
+
+  type 'a t = { mask : int; buckets : 'a Llist.marked Atomic.t array }
+
+  let init ~size_exponent =
+    let size = Int.shift_left 1 size_exponent in
+    let mask = size - 1 in
+    let llist = Llist.init () in
+    {
+      mask;
+      buckets =
+        Array.init size (fun i ->
+            if i = 0 then Atomic.make (new_dummy_node 0 llist)
+            else Atomic.make Llist.Last);
+    }
+
+  let rec init_bucket buckets ind =
+    let parent_ind = unset_msb ind in
+    let parent_bucket = buckets.(parent_ind) in
+
+    (if parent_ind <> ind then
+     match Atomic.get parent_bucket with
+     | Llist.Last -> init_bucket buckets parent_ind
+     | LRemove -> failwith "Should never happen."
+     | _ -> ());
+
+    new_dummy_node ind parent_bucket |> Atomic.set buckets.(ind)
+
+  (** [get_bucket_id key buckets mask] searches the bucket's index corresponding
+      to [key] (= [key mod t.size]), initializes if needed and returns
+      it. *)
+  let get_bucket_ind key buckets mask =
+    let ind = key land mask in
+    let bucket = buckets.(ind) in
+    match Atomic.get bucket with
+    | Llist.Last ->
+        init_bucket buckets ind;
+        buckets.(ind)
+    | LRemove -> failwith "Should never happen."
+    | _ -> bucket
+
+  (** [add key value t] *)
+  let add key value { buckets; mask; _ } =
+    Llist.add (compute_hkey key) (Regular value)
+    @@ get_bucket_ind key buckets mask
+
+  let find key { buckets; mask; _ } =
+    let is_found, local =
+      Llist.find (compute_hkey key) (get_bucket_ind key buckets mask)
+    in
+    if not is_found then None
+    else
+      match local.curr with
+      | Normal { value = Regular k; _ } | Remove { value = Regular k; _ } ->
+          Some k
```

Comment:
In the `Remove {..}` case the node is logically removed and is technically not in the linked list.

Reply:
(The hashtable is section 13.3 from The Art of Multiprocessor Programming.)

Reply:
The paper I worked with is actually _Split-ordered lists: Lock-free extensible hash tables_ from O. Shalev and N. Shavit. I realized only a few weeks ago that it was better described and improved in _The Art of Multiprocessor Programming_. 

I will add the simplest algorithm improvements in this PR and will add the other ones in a later PR.  

The ones that I will include soon are : 
- (sure) wait-free `mem` function (may be keep both : a wait-free one and the current one that is helping with removal as they may be both useful)  
- (probably) removal loop in the `Llist.find` function as written in the `LockfreeList` class in the AoMP book
- (probably) changing the criteria to determine when to grow to reduce contention (currently every add and remove change a single value).

In another PR ; currently this hash table is actually extensible and not really resizable. It can only grow and has a maximum size. There are ideas on the book about how to make it resizable without limit (not sure about shrinking though).